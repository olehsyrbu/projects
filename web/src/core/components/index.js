export * from './Alert';
export * from './Avatar';
export * from './Dialog';
export * from './Dot';
export * from './Badge';
export * from './Favorite';
export * from './DateRangePicker';
export * from './DatePicker';
export * from './Toolbar';
export * from './Switch';
export * from './InlineSelect';
export * from './Radio';
export * from './Menu';
