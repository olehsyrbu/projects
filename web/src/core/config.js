import logoUrl from '@/images/logo.svg';

export function readConfigs(overrides) {
  let sentryTraceSampleRate = process.env.MIR_SENTRY_TRACE_SAMPLE_RATE;
  let sentryDebug = process.env.MIR_SENTRY_DEBUG;

  if (process.env.NODE_ENV === 'production') {
    sentryTraceSampleRate = overrides.monitoring.sentryTraceSampleRate;
    sentryDebug = overrides.monitoring.sentryDebug;
  }

  sentryTraceSampleRate = Number.parseFloat(sentryTraceSampleRate) || 0;
  sentryDebug = sentryDebug === 'true';

  const defaults = {
    deploymentUrl: process.env.MIR_DEPLOYMENT_URL,
    googleMapsApiKey: process.env.MIR_GOOGLE_MAP_API_KEY,
    mixpanelToken: process.env.MIR_MIXPANEL_TOKEN,
    zendeskKey: process.env.MIR_ZENDESK_KEY,
    fullstoryOrgId: process.env.MIR_FULLSTORY_ORG_ID,
    copyrightPolicyUrl: process.env.MIR_COPYRIGHT_POLICY,
    privacyPolicyUrl: process.env.MIR_PRIVACY_POLICY,
    termsOfUseUrl: process.env.MIR_TERMS_OF_USE,
    api: {
      baseUrl: process.env.MIR_API_GATEWAY_BASE_URL || 'http://localhost:3000/graphql',
    },
    // hardcoding mapping of orgs to insurance types, we are going to migrate to backend when more insurances come in
    organizationInsuranceTypes: {
      bcbsks: ['BCBS04'],
    },
    whiteLabeling: {
      mainLogoUrl: logoUrl,
      orgName: '',
      colors: [],
      subdomain: process.env.MIR_DOMAIN_DEV_ORG_FALLBACK,
    },
    rootOrganization: {
      slug: 'miresource',
      name: 'MiResource',
    },
    auth: {
      refreshSleepTimeout: process.env.MIR_AUTH_REFRESH_SLEEP_TIMEOUT_SEC || 180,
      userIdleTimeout: process.env.MIR_AUTH_USER_IDLE_TIMEOUT_SEC || 1750,
    },
    monitoring: {
      sentryDsn: process.env.MIR_SENTRY_DSN,
      sentryEnv: process.env.MIR_SENTRY_ENV,
      sentryTraceSampleRate,
      sentryDebug,
    },
    launchDarklyClientId: process.env.MIR_DARK_LAUNCH_CLIENT_ID,
    shouldMockNetwork:
      process.env.MIR_MOCK_NETWORK === 'true' && process.env.NODE_ENV === 'development',
    onDemandCaseManagementUrl: process.env.MIR_ON_DEMAND_CASE_MANAGEMENT_URL,
    mindLogger: {
      baseUrl: process.env.MIR_MINDLOGGER_API_URL,
      credentials: process.env.MIR_MINDLOGGER_API_CREDENTIALS,
    },
    learningCenterUrl: 'https://www.learn.miresource.com',
    gpaUploadTemplate: 'https://assets.miresource.com/GPA_template.xlsx',
    ncsuEmergencyUrl: 'https://assets.miresource.com/NCSU_mental_health_emergency.pdf',
  };

  let result = defaults;

  if (process.env.NODE_ENV === 'production') {
    result = {
      ...defaults,
      ...overrides,
      monitoring: {
        ...overrides.monitoring,
        sentryTraceSampleRate,
        sentryDebug,
      },
    };
  }

  return result;
}

export default readConfigs(window.__MIR_APP__ ? window.__MIR_APP__.config : {});
