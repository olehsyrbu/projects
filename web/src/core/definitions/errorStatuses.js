export const ERROR_STATUSES = {
  CONFLICT: 409,
};
