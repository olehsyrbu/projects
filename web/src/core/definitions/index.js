export * from './invitations';
export * from './roleConstants';
export * from './referenceDataConstants';
export * from './errorStatuses';
