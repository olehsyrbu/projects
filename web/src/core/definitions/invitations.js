export const InvitationTypes = {
  ReferralCoordinator: 'REFERRAL_COORDINATOR',
  OrganizationMember: 'ORGANIZATION_MEMBER',
  Provider: 'PROVIDER',
  NETWORK_MANAGER: 'NETWORK_MANAGER',
};

export const InvitationStatus = {
  Open: 'OPEN',
  Accepted: 'ACCEPTED',
  Declined: 'DECLINED',
  Withdrawn: 'WITHDRAWN',
};
