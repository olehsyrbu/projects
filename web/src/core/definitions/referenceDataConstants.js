export const AVAILABILITY_TYPES = {
  ACCEPTING_NEW_CLIENTS: 'ACCEPTING_NEW_CLIENTS',
  NOT_ACCEPTING_NEW_CLIENTS: 'NOT_ACCEPTING_NEW_CLIENTS',
  PLEASE_INQUIRE: 'PLEASE_INQUIRE',
};
