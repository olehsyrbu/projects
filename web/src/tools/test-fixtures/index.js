export * from './invitationFixtures';
export * from './userProfileFixtures';
export * from './userFixtures';
export * from './organizationFixtures';
export * from './abilitiesFixture';
export * from './providerFixtures';
export * from './teamNotesFixtures';
export * from './pageInfoFixtures';
