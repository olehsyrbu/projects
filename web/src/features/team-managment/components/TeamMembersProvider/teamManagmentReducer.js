import { teamMemberSpecification } from '@/features/team-managment/utils';
import { InvitationTypes, ROLE_NAMES } from '@/core/definitions';
import { cloneDeep } from 'lodash-es';

export const initState = {
  admins: [],
  members: [],
  userId: '',
};

export const TeamManagementActions = {
  INIT: 'INIT',
  REMOVE_MEMBER: 'REMOVE_MEMBER',
  SWITCH_ROLE: 'SWITCH_ROLE',
};

export function teamManagementReducer(state = initState, action) {
  let admins, members;
  switch (action.type) {
    case TeamManagementActions.INIT:
      const { invitations, currentUserId } = action.payload;
      const result = invitations.reduce(
        (memo, invitation) => {
          let user = invitation.user;
          let isAdmin;

          if (user) {
            isAdmin = user && teamMemberSpecification.isTeamAdmin(user);
          } else {
            isAdmin = invitation.type === InvitationTypes.ReferralCoordinator;
          }

          if (isAdmin) {
            memo.admins.push(invitation);
          } else {
            memo.members.push(invitation);
          }

          return memo;
        },
        {
          admins: [],
          members: [],
        },
      );

      const findByUserId = currentUserById(currentUserId, 'equal');
      const excludeByUserId = currentUserById(currentUserId);

      return {
        ...initState,
        userId: currentUserId,
        admins: [
          ...result.admins.filter(findByUserId),
          ...result.admins.filter(excludeByUserId).sort(sortBySentAt),
        ],
        members: result.members.sort(sortBySentAt),
      };

    case TeamManagementActions.REMOVE_MEMBER:
      let byId = excludeById(action.payload.invitationId);

      return {
        ...state,
        admins: state.admins.filter(byId),
        members: state.members.filter(byId),
      };

    case TeamManagementActions.SWITCH_ROLE:
      const { invitationId } = action.payload;

      members = state.members;
      admins = state.admins;

      const includeInviteAndHasAUser = ({ id, user }) => id === invitationId && user;
      const byInvitationId = (item) => item.id === invitationId;

      if (members.some(includeInviteAndHasAUser)) {
        admins = cloneDeep(state.admins);

        let member = cloneDeep(members.find(byInvitationId));
        members = members.filter(excludeById(invitationId));
        let user = member.user;
        user.role = ROLE_NAMES.REFERRAL_COORDINATOR;

        admins.push(member);

        const findByUserId = currentUserById(state.userId, 'equal');
        const excludeByUserId = currentUserById(state.userId);

        return {
          ...state,
          admins: [
            ...admins.filter(findByUserId),
            ...admins.filter(excludeByUserId).sort(sortBySentAt),
          ],
          members,
        };
      } else if (admins.some(includeInviteAndHasAUser)) {
        members = cloneDeep(members);

        let admin = cloneDeep(admins.find(byInvitationId));
        admins = admins.filter(excludeById(invitationId));
        let user = admin.user;
        user.role = ROLE_NAMES.TEAM_MEMBER;

        members.push(admin);
        members.sort(sortBySentAt);

        return {
          ...state,
          admins,
          members,
        };
      }

      return {
        ...state,
      };

    default:
      return { ...state };
  }
}

const excludeById = (id) => (invitation) => {
  return invitation.id !== id;
};

function sortBySentAt(left, right) {
  return new Date(right.sentAt) - new Date(left.sentAt);
}

const currentUserById =
  (userId, operation) =>
  ({ user }) => {
    return operation === 'equal' ? user?.id === userId : user?.id !== userId;
  };
