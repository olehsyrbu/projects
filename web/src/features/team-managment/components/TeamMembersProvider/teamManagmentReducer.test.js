import { initState, TeamManagementActions, teamManagementReducer } from './teamManagmentReducer';
import { invitationFixture } from '@/tools/test-fixtures';
import { InvitationStatus, InvitationTypes } from '@/core/definitions';

function getInitialData() {
  const firstTeamAdmin = invitationFixture.teamMember({
    id: 'firstTeamAdmin',
    firstName: 'Aydon',
    lastName: 'Carter',
    email: 'john_k@familypractice.com',
    isAdmin: true,
    sentAt: '2021-11-25T07:10:19.697Z',
  });

  const secondTeamAdmin = invitationFixture.teamMember({
    id: 'secondTeamAdmin',
    firstName: 'Bngeborga',
    lastName: 'Smedes',
    email: 'i.smedes@gmail.com',
    isAdmin: true,
    sentAt: '2021-11-23T07:10:19.697Z',
  });

  const firstTeamMember = invitationFixture.teamMember({
    id: 'firstTeamMember',
    firstName: 'Anton',
    lastName: 'Joe',
    email: 'joe@example.com',
    isAdmin: false,
    sentAt: '2021-11-25T07:10:19.697Z',
  });

  const secondTeamMember = invitationFixture.teamMember({
    id: 'secondTeamMember',
    firstName: 'Bill',
    lastName: 'Tomson',
    email: 'bill@example.com',
    isAdmin: false,
    user: {
      id: '4',
    },
    sentAt: '2021-11-23T07:10:19.697Z',
  });

  const thirdTeamAdmin = invitationFixture.invitationWithoutUser({
    id: 'thirdTeamAdmin',
    firstName: 'Yusy',
    lastName: 'Boo',
    email: 'foo@email.com',
    status: InvitationStatus.Open,
    type: InvitationTypes.ReferralCoordinator,
    sentAt: '2021-11-20T07:10:19.697Z',
  });

  return {
    thirdTeamAdmin,
    secondTeamMember,
    firstTeamAdmin,
    firstTeamMember,
    secondTeamAdmin,
  };
}

it('provides default', () => {
  const actual = teamManagementReducer(initState, {});

  expect(actual).toEqual(initState);
});

it('inti with new data and split to members and admins', () => {
  const { thirdTeamAdmin, secondTeamMember, firstTeamAdmin, firstTeamMember, secondTeamAdmin } =
    getInitialData();

  const actual = teamManagementReducer(initState, {
    type: TeamManagementActions.INIT,
    payload: {
      invitations: [
        thirdTeamAdmin,
        secondTeamMember,
        firstTeamAdmin,
        firstTeamMember,
        secondTeamAdmin,
      ],
      currentUserId: 'user-secondTeamAdmin',
    },
  });

  expect(actual).toEqual({
    members: [firstTeamMember, secondTeamMember],
    admins: [secondTeamAdmin, firstTeamAdmin, thirdTeamAdmin],
    userId: 'user-secondTeamAdmin',
  });
});

it('remove admin', () => {
  const { firstTeamAdmin, firstTeamMember, secondTeamAdmin } = getInitialData();

  let actual = teamManagementReducer(null, {
    type: TeamManagementActions.INIT,
    payload: {
      invitations: [firstTeamAdmin, firstTeamMember, secondTeamAdmin],
    },
  });

  actual = teamManagementReducer(actual, {
    type: TeamManagementActions.REMOVE_MEMBER,
    payload: {
      invitationId: secondTeamAdmin.id,
    },
  });

  expect(actual).toMatchSnapshot();
});

it('remove member', () => {
  const { firstTeamAdmin, firstTeamMember, secondTeamAdmin } = getInitialData();

  let actual = teamManagementReducer(null, {
    type: TeamManagementActions.INIT,
    payload: {
      invitations: [firstTeamAdmin, firstTeamMember, secondTeamAdmin],
    },
  });

  actual = teamManagementReducer(actual, {
    type: TeamManagementActions.REMOVE_MEMBER,
    payload: {
      invitationId: firstTeamMember.id,
    },
  });

  expect(actual).toMatchSnapshot();
});

it('promote member to admin', async () => {
  const { firstTeamAdmin, firstTeamMember, secondTeamMember } = getInitialData();

  let actual = teamManagementReducer(null, {
    type: TeamManagementActions.INIT,
    payload: {
      invitations: [firstTeamAdmin, firstTeamMember, secondTeamMember],
    },
  });

  actual = teamManagementReducer(actual, {
    type: TeamManagementActions.SWITCH_ROLE,
    payload: {
      invitationId: secondTeamMember.id,
      isAdmin: true,
    },
  });

  expect(actual).toMatchSnapshot();
});

it('downgrade admin to member', async () => {
  const { firstTeamAdmin, firstTeamMember, secondTeamAdmin } = getInitialData();

  let actual = teamManagementReducer(null, {
    type: TeamManagementActions.INIT,
    payload: {
      invitations: [firstTeamAdmin, firstTeamMember, secondTeamAdmin],
    },
  });

  actual = teamManagementReducer(actual, {
    type: TeamManagementActions.SWITCH_ROLE,
    payload: {
      invitationId: secondTeamAdmin.id,
    },
  });

  expect(actual).toMatchSnapshot();
});
