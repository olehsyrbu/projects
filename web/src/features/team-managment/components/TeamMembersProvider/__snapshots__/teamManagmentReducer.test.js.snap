// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downgrade admin to member 1`] = `
{
  "admins": [
    {
      "email": "john_k@familypractice.com",
      "id": "firstTeamAdmin",
      "invitationCode": "code-firstTeamAdmin",
      "messageType": "EMAIL",
      "recipientName": "Aydon Carter",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "john_k@familypractice.com",
        "firstName": "Aydon",
        "id": "user-firstTeamAdmin",
        "lastName": "Carter",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "REFERRAL_COORDINATOR",
      },
    },
  ],
  "members": [
    {
      "email": "joe@example.com",
      "id": "firstTeamMember",
      "invitationCode": "code-firstTeamMember",
      "messageType": "EMAIL",
      "recipientName": "Anton Joe",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "joe@example.com",
        "firstName": "Anton",
        "id": "user-firstTeamMember",
        "lastName": "Joe",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "TEAM_MEMBER",
      },
    },
    {
      "email": "i.smedes@gmail.com",
      "id": "secondTeamAdmin",
      "invitationCode": "code-secondTeamAdmin",
      "messageType": "EMAIL",
      "recipientName": "Bngeborga Smedes",
      "sentAt": "2021-11-23T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "i.smedes@gmail.com",
        "firstName": "Bngeborga",
        "id": "user-secondTeamAdmin",
        "lastName": "Smedes",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "TEAM_MEMBER",
      },
    },
  ],
  "userId": undefined,
}
`;

exports[`promote member to admin 1`] = `
{
  "admins": [
    {
      "email": "john_k@familypractice.com",
      "id": "firstTeamAdmin",
      "invitationCode": "code-firstTeamAdmin",
      "messageType": "EMAIL",
      "recipientName": "Aydon Carter",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "john_k@familypractice.com",
        "firstName": "Aydon",
        "id": "user-firstTeamAdmin",
        "lastName": "Carter",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "REFERRAL_COORDINATOR",
      },
    },
    {
      "email": "bill@example.com",
      "id": "secondTeamMember",
      "invitationCode": "code-secondTeamMember",
      "messageType": "EMAIL",
      "recipientName": "Bill Tomson",
      "sentAt": "2021-11-23T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "bill@example.com",
        "firstName": "Bill",
        "id": "user-secondTeamMember",
        "lastName": "Tomson",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "REFERRAL_COORDINATOR",
      },
    },
  ],
  "members": [
    {
      "email": "joe@example.com",
      "id": "firstTeamMember",
      "invitationCode": "code-firstTeamMember",
      "messageType": "EMAIL",
      "recipientName": "Anton Joe",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "joe@example.com",
        "firstName": "Anton",
        "id": "user-firstTeamMember",
        "lastName": "Joe",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "TEAM_MEMBER",
      },
    },
  ],
  "userId": undefined,
}
`;

exports[`remove admin 1`] = `
{
  "admins": [
    {
      "email": "john_k@familypractice.com",
      "id": "firstTeamAdmin",
      "invitationCode": "code-firstTeamAdmin",
      "messageType": "EMAIL",
      "recipientName": "Aydon Carter",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "john_k@familypractice.com",
        "firstName": "Aydon",
        "id": "user-firstTeamAdmin",
        "lastName": "Carter",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "REFERRAL_COORDINATOR",
      },
    },
  ],
  "members": [
    {
      "email": "joe@example.com",
      "id": "firstTeamMember",
      "invitationCode": "code-firstTeamMember",
      "messageType": "EMAIL",
      "recipientName": "Anton Joe",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "joe@example.com",
        "firstName": "Anton",
        "id": "user-firstTeamMember",
        "lastName": "Joe",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "TEAM_MEMBER",
      },
    },
  ],
  "userId": undefined,
}
`;

exports[`remove member 1`] = `
{
  "admins": [
    {
      "email": "john_k@familypractice.com",
      "id": "firstTeamAdmin",
      "invitationCode": "code-firstTeamAdmin",
      "messageType": "EMAIL",
      "recipientName": "Aydon Carter",
      "sentAt": "2021-11-25T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "john_k@familypractice.com",
        "firstName": "Aydon",
        "id": "user-firstTeamAdmin",
        "lastName": "Carter",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "REFERRAL_COORDINATOR",
      },
    },
    {
      "email": "i.smedes@gmail.com",
      "id": "secondTeamAdmin",
      "invitationCode": "code-secondTeamAdmin",
      "messageType": "EMAIL",
      "recipientName": "Bngeborga Smedes",
      "sentAt": "2021-11-23T07:10:19.697Z",
      "status": "OPEN",
      "type": "ORGANIZATION_MEMBER",
      "user": {
        "email": "i.smedes@gmail.com",
        "firstName": "Bngeborga",
        "id": "user-secondTeamAdmin",
        "lastName": "Smedes",
        "loggedInAt": "2021-11-05T11:57:37.686Z",
        "role": "REFERRAL_COORDINATOR",
      },
    },
  ],
  "members": [],
  "userId": undefined,
}
`;
